framework/CodeInterfaces/RAVEN/RAVENparser.py

```diff
@@ -0,0 +1,295 @@
+          changeTheNode = False
+        else:
+          changeTheNode = True
+      else:
+        # there are no attributes that are needed to track down the node to change
+        pathNode = './/' + node.replace("|","/").strip()
+        if allowAdd:
+          pathNodeTemp = './/'
+          for component in node.replace("|","/").strip():
+            pathNodeTemp += component
+            if len(returnElement.findall(pathNodeTemp)) > 0:
+              allowAddNodes.append(pathNodeTemp)
+            else:
+              allowAddNodes.append(None)
+            #allowAddNodes.append(returnElement.findall(pathNodeTemp))
+            allowAddNodesPath[component.strip()] = None
+      # look for the node with XPath directives
+      foundNodes = returnElement.findall(pathNode)
+      if len(foundNodes) > 1:
+        raise IOError(self.printTag+' ERROR: multiple nodes have been found corresponding to path -> '+node.strip()+'. Please use the attribute identifier "@" to nail down to a specific node !!')
+      if len(foundNodes) == 0 and not allowAdd:
+        raise IOError(self.printTag+' ERROR: no node has been found corresponding to path -> '+node.strip()+'. Please check the input!!')
+      if len(foundNodes) == 0:
+        # this means that the allowAdd is true (=> no error message has been raised)
+        indexFirstUnknownNode = allowAddNodes.index(None)
+        getFirstElement = returnElement.findall(allowAddNodes[indexFirstUnknownNode-1])[0]
+        for i in range(indexFirstUnknownNode,len(allowAddNodes)):
+          nodeWithAttributeName = allowAddNodesPath.keys()[i]
+          if not allowAddNodesPath[nodeWithAttributeName]:
+            subElement =  ET.Element(nodeWithAttributeName)
+          else:
+            subElement =  ET.Element(nodeWithAttributeName, attrib=allowAddNodesPath[nodeWithAttributeName])
+          getFirstElement.append(subElement)
+          getFirstElement = subElement
+        if changeTheNode:
+          subElement.text = str(value).strip()
+        else:
+          subElement.attrib[attribConstruct.keys()[-1]] = str(value).strip()
+
+      else:
+        nodeToChange = foundNodes[0]
+        pathNode     = './/'
+        if changeTheNode:
+          nodeToChange.text = str(value).strip()
+        else:
+          nodeToChange.attrib[attribName] = str(value).strip()
+    return returnElement
+
+if __name__ == '__main__':
+  parser = RAVENparser("/Users/alfoa/projects/raven_github/raven/tests/framework/test_redundant_inputs.xml")
+  modifyDict = {'Distributions|Uniform@name:a_dist|lowerBound':-1,'RunInfo|batchSize':2,'Models|Code@name:py_script@subType:GenericCode|prepend':"Python"}
+  lupo = parser.modifyOrAdd(modifyDict,allowAdd=False)
+  modifyDict['Distributions|Normal@name:Andrea|mean'] = 22222
+  modifyDict['Distributions|Normal@name:Andrea|sigma'] = 2
+  parser.printInput(lupo,outfile="lupo1.xml")
+  lupo2 = parser.modifyOrAdd(modifyDict,allowAdd=True)
+  parser.printInput(lupo2,outfile="lupo2.xml")
```

Comment:
I like the unit test, but it seems to only apply to your user directory.

Reply:
right....leftover... I will modify it.